dot/core/messages.go

```diff
@@ -20,24 +20,30 @@ func (s *Service) validateTransaction(head *types.Header, rt RuntimeInstance,
 	s.storageState.Lock()
 
 	ts, err := s.storageState.TrieState(&head.StateRoot)
-	s.storageState.Unlock()
 	if err != nil {
-		return nil, fmt.Errorf("cannot get trie state from storage for root %s: %w", head.StateRoot, err)
+		s.storageState.Unlock()
+		return nil, fmt.Errorf("getting trie state from storage: %w", err)
 	}
 
-	rt.SetContextStorage(ts)
+	// ValidateTransaction modifies the trie state so we want to snapshot it
+	// so that the original trie state remains unaffected.
+	temporaryState := ts.Snapshot()
+
+	s.storageState.Unlock()
 
 	// validate each transaction
 	externalExt, err := s.buildExternalTransaction(rt, tx)
 	if err != nil {
 		return nil, fmt.Errorf("building external transaction: %w", err)
 	}
 
+	rt.SetContextStorage(temporaryState)
 	validity, err = rt.ValidateTransaction(externalExt)
 	if err != nil {
 		logger.Debugf("failed to validate transaction: %s", err)
 		return nil, err
 	}
+	rt.SetContextStorage(nil)
```

Comment:
why do we have to do this now?

Reply:
Clearing the storage trie state for that runtime, so that it can be garbage collected. Otherwise the runtime may hold it forever.
The point is to reduce/avoid memory leaks, so this aims at setting the state and then clearing it from the runtime structure; all calls get a new trie state anyway